File: 2026/06-June/03-Wednesday/solution.py
from bisect import bisect_left

class Solution:
    def earliestFinishTime(self,
                           landStartTime,
                           landDuration,
                           waterStartTime,
                           waterDuration):

        def solve(s1, d1, s2, d2):

            rides = sorted(zip(s2, d2))

            start = [x[0] for x in rides]

            m = len(rides)

            pref = [0] * m
            suf = [0] * m

            pref[0] = rides[0][1]

            for i in range(1, m):
                pref[i] = min(pref[i - 1], rides[i][1])

            suf[-1] = rides[-1][0] + rides[-1][1]

            for i in range(m - 2, -1, -1):
                suf[i] = min(
                    suf[i + 1],
                    rides[i][0] + rides[i][1]
                )

            ans = float("inf")

            for st, dur in zip(s1, d1):

                finish = st + dur

                pos = bisect_left(start, finish)

                if pos:
                    ans = min(ans,
                              finish + pref[pos - 1])

                if pos < m:
                    ans = min(ans,
                              suf[pos])

            return ans

        return min(
            solve(landStartTime, landDuration,
                  waterStartTime, waterDuration),
            solve(waterStartTime, waterDuration,
                  landStartTime, landDuration)
        )

File: 2026/06-June/03-Wednesday/solution.js
var earliestFinishTime = function(
    landStartTime,
    landDuration,
    waterStartTime,
    waterDuration
) {

    const solve = (s1, d1, s2, d2) => {

        let rides = [];

        for (let i = 0; i < s2.length; i++)
            rides.push([s2[i], d2[i]]);

        rides.sort((a, b) => a[0] - b[0]);

        const m = rides.length;

        const start = Array(m);
        const pref = Array(m);
        const suf = Array(m);

        for (let i = 0; i < m; i++)
            start[i] = rides[i][0];

        pref[0] = rides[0][1];

        for (let i = 1; i < m; i++)
            pref[i] = Math.min(pref[i - 1], rides[i][1]);

        suf[m - 1] = rides[m - 1][0] + rides[m - 1][1];

        for (let i = m - 2; i >= 0; i--)
            suf[i] = Math.min(
                suf[i + 1],
                rides[i][0] + rides[i][1]
            );

        let ans = Number.MAX_SAFE_INTEGER;

        for (let i = 0; i < s1.length; i++) {

            let finish = s1[i] + d1[i];

            let l = 0, r = m;

            while (l < r) {
                let mid = (l + r) >> 1;
                if (start[mid] < finish)
                    l = mid + 1;
                else
                    r = mid;
            }

            let pos = l;

            if (pos > 0)
                ans = Math.min(
                    ans,
                    finish + pref[pos - 1]
                );

            if (pos < m)
                ans = Math.min(
                    ans,
                    suf[pos]
                );
        }

        return ans;
    };

    return Math.min(
        solve(landStartTime, landDuration,
              waterStartTime, waterDuration),
        solve(waterStartTime, waterDuration,
              landStartTime, landDuration)
    );
};
